File: CS50 problem sets/pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main()
{
    int sum1, sum2, firstNum, secNum, twoDigits, checkSum, firstTowDigit, numOfDigits, i, j;
    long cardNum, vCardNum;
    
    do
    {   
        //enter credti card number
        cardNum = get_long("Put your credit card number: ");
    }
    while (cardNum <= 0);
    
    
    vCardNum = cardNum;
    firstNum = 0;
    secNum = 0;
    sum1 = 0;
    sum2 = 0;
    i = 0;
    
    // sum from last-to-second digit
    while (cardNum > 0)
    {
        firstNum = cardNum % 10;
        sum1 = sum1 + firstNum;
        cardNum = cardNum / 100 ;
        // count the digits
        i++;
    }
    
    j = 0;

    vCardNum = vCardNum / 10;
     
    //multiply and sum befor the last-to-first digit
    while (vCardNum > 0)
    {
        secNum = vCardNum % 10;
         
        twoDigits = secNum * 2;
        //int n = twoDigits / 10;
        sum2 = sum2 + (twoDigits % 10) + (twoDigits / 10);
         
        vCardNum = vCardNum / 100;
        // count the digits
        j++; 
         
    }
     
    //sum the two first numbers in the card
    firstTowDigit = secNum + firstNum; 
     
    //check if it's vaild or not
    checkSum = sum1 + sum2; 
     
    checkSum = checkSum % 10;
     
    // catculate the number of card digits
    numOfDigits = i + j;
    printf("%d \n",sum1);
    printf("%d \n",sum2 );
    printf("%d \n",numOfDigits );
    printf("%d \n",checkSum);
    
     
    //check the type of the credit card
    if (checkSum == 0 && (twoDigits == 4 || secNum == 4)  && (numOfDigits == 13 || numOfDigits == 16))
    {
        printf("VISA\n");
    }
    else if (checkSum == 0 && numOfDigits == 16 && (firstTowDigit >= 6 && firstTowDigit <= 10))
    {
        printf("MASTERCARD\n");
    }
    else if (checkSum == 0 && numOfDigits == 15 && (firstTowDigit == 7 || firstTowDigit == 10))
    {
        printf("AMEX\n");
    }
    else 
    {
        printf("INVALID\n");
    }
}

File: CS50 problem sets/pset6/credit/credit.py
from cs50 import get_string
import re


def main():
    # asking the user to enter the credit card number
    while True:
        card_num = get_string("Put your credit card number: ")
        # check if credit card number in valid or not 
        pattern = '[0-9]{13,16}'
        x = re.match(pattern, card_num)
        if x:
            break
        print("INVALID")
        break

    card_num2 = int(card_num)
    Vcard_num = card_num2
    first_num = 0
    sec_num = 0
    sum1 = 0
    sum2 = 0
    sum3 = 0
    i = 0
    
    # loop on the numbers to sum the numbers beginning from the first
    while card_num2 > 0:
        first_num = int(card_num2 % 10)
        sum3 += first_num
        card_num2 = int(card_num2 / 100)
        i += 1
        
    j = 0
    Vcard_num = Vcard_num / 10
    # loop through the numbers beginning from the second number
    while (Vcard_num > 0):
        sec_num = int(Vcard_num % 10) 
        two_digits = (sec_num * 2)
        sum3 = sum3 + int(two_digits % 10) + int(two_digits / 10)
        Vcard_num = int(Vcard_num / 100)
        j += 1

    first_two_digits = sec_num + first_num
    
    sum3 = sum3 % 10
    num_of_digits = i + j
   # check the type of the credit card according to thier system
    if sum3 == 0 and (two_digits == 4 or sec_num == 4) and (num_of_digits == 13 or num_of_digits == 16):
        print("VISA")
    elif sum3 == 0 and num_of_digits == 16 and (6 <= first_two_digits <= 10):
        print("MASTERCARD")
    elif sum3 == 0 and num_of_digits == 15 and (first_two_digits == 7 or 10):
        print("AMEX")
 
  
if __name__ == "__main__":
    main()